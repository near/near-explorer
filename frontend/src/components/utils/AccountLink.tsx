import Link from "next/link";

export interface Props {
  accountId: string;
  children?: React.ReactNode;
}

export default ({ accountId, children }: Props) => (
  <Link href="/accounts/[id]" as={`/accounts/${accountId}`}>
    <a className="account-link">{children || `@${accountId}`}</a>
  </Link>
);
