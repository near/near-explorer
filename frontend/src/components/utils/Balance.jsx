/// Copied from near-wallet project:
/// https://github.com/nearprotocol/near-wallet/blob/41cb65246134308dd553b532dfb314b45b38b65c/src/components/common/Balance.js

import { utils } from "nearlib";
import { OverlayTrigger, Tooltip } from "react-bootstrap";

const Balance = ({ amount }) => {
  if (!amount) {
    throw new Error("amount property should not be null");
  }
  let amountShow = convertToShow(amount);
  let amountPrecise = formatPreciseNEAR(amount);
  return (
    <OverlayTrigger
      placement={"bottom"}
      overlay={<Tooltip>{amountPrecise} yoctoⓃ</Tooltip>}
    >
      <span>{amountShow} Ⓝ</span>
    </OverlayTrigger>
  );
};

const convertToShow = amount => {
  if (amount === "0") {
    return "0";
  }
  return formatNEAR(amount);
};

const formatNEAR = amount => {
  let ret = utils.format.formatNearAmount(amount, 5);
  if (ret === "0" && amount > 0) {
    return "<0.00001";
  }
  return ret;
};

const formatPreciseNEAR = amount => {
  const REG = /(?=(\B)(\d{3})+$)/g;
  return amount.replace(REG, ",");
};
export default Balance;
