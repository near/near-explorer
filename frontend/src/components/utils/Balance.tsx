import JSBI from "jsbi";
import * as React from "react";
import { OverlayTrigger, Tooltip } from "react-bootstrap";
import * as BI from "../../libraries/bigint";

interface Props {
  amount: string | JSBI;
  label?: React.ReactNode;
  suffix?: React.ReactNode;
  className?: string;
  formulatedAmount?: string;
  fracDigits?: number;
}

const Balance: React.FC<Props> = React.memo(
  ({
    amount,
    label = null,
    suffix = undefined,
    className,
    formulatedAmount = undefined,
    fracDigits = 5,
  }) => {
    if (!amount) {
      throw new Error("amount property should not be null");
    }

    const defaultLabel = "Ⓝ";

    let amountShow = !formulatedAmount
      ? formatNEAR(amount, fracDigits)
      : formulatedAmount;
    let amountPrecise = showInYocto(amount.toString());
    return (
      <OverlayTrigger
        placement={"bottom"}
        overlay={<Tooltip id="balance">{amountPrecise}</Tooltip>}
      >
        <span className={className}>
          {amountShow}
          {suffix}
          &nbsp;
          {label ?? defaultLabel}
        </span>
      </OverlayTrigger>
    );
  }
);

const ROUNDING_OFFSETS: JSBI[] = [];
for (
  let i = 0, offset = JSBI.BigInt(5);
  i < BI.nearNominationExponent;
  i++, offset = JSBI.multiply(offset, BI.ten)
) {
  ROUNDING_OFFSETS[i] = offset;
}

const formatNearAmount = (
  balance: string | JSBI,
  fracDigits: number
): string => {
  let balanceBN = JSBI.BigInt(balance);
  if (fracDigits !== BI.nearNominationExponent) {
    // Adjust balance for rounding at given number of digits
    const roundingExp = BI.nearNominationExponent - fracDigits - 1;
    if (roundingExp > 0) {
      balanceBN = JSBI.add(balanceBN, ROUNDING_OFFSETS[roundingExp]);
    }
  }
  balance = balanceBN.toString();
  const wholeStr =
    balance.substring(0, balance.length - BI.nearNominationExponent) || "0";
  const fractionStr = balance
    .substring(balance.length - BI.nearNominationExponent)
    .padStart(BI.nearNominationExponent, "0")
    .substring(0, fracDigits);
  return `${formatWithCommas(wholeStr)}.${fractionStr}`.replace(/\.?0*$/, "");
};

export const formatNEAR = (amount: string | JSBI, fracDigits = 5): string => {
  const ret = formatNearAmount(amount.toString(), fracDigits);

  if (amount === "0") {
    return amount;
  } else if (ret === "0") {
    return `<${!fracDigits ? `0` : `0.${"0".repeat((fracDigits || 1) - 1)}1`}`;
  }
  return ret;
};

export const showInYocto = (amountStr: string) => {
  return formatWithCommas(amountStr) + " yoctoⓃ";
};

export const formatWithCommas = (value: string): string => {
  const pattern = /(-?\d+)(\d{3})/;
  while (pattern.test(value)) {
    value = value.toString().replace(pattern, "$1,$2");
  }
  return value;
};

export default Balance;
