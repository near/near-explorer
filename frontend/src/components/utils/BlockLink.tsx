import Link from "next/link";

export interface Props {
  blockHash: string;
  children?: React.ReactNode;
}

export default ({ blockHash, children }: Props) => (
  <Link href="/blocks/[hash]" as={`/blocks/${blockHash}`}>
    <a className="block-link">
      {children || `${blockHash.substring(0, 7)}...`}
    </a>
  </Link>
);
