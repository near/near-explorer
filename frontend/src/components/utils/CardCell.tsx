import { Row, Col, Card } from "react-bootstrap";

export interface Props {
  className?: string;
  title: React.ReactElement | string;
  imgLink?: string;
  text: React.ReactElement | string;
}

export default ({ title, imgLink, text, className }: Props) => (
  <Card className={`card-cell ${className || ""}`}>
    <Card.Body>
      <Row noGutters>
        <Col xs="auto" md="12" className="card-cell-title align-self-center">
          {imgLink && <img src={imgLink} className="card-cell-title-img" />}
          {title}
        </Col>
        <Col
          xs="auto"
          md="12"
          className="ml-auto card-cell-text align-self-center"
        >
          {text}
        </Col>
      </Row>
    </Card.Body>
    <style jsx global>{`
      .card-cell {
        border-style: solid;
        border-color: #e6e6e6;
        border-width: 0 0 0 2px;
        border-radius: 0;
      }

      .card-cell-title {
        text-transform: uppercase;
        letter-spacing: 1.8px;
        color: #999999;
        font-family: BentonSans;
        font-size: 14px;
        font-weight: 500;
      }

      .card-cell-title-img {
        width: 12px !important;
        margin-right: 8px;
        margin-top: -3px;
      }

      .card-cell-text {
        font-family: BwSeidoRound;
        font-size: 18px;
        font-weight: 500;
        color: #24272a;
      }

      .card-cell a {
        font-weight: 500;
        color: #6ad1e3;
        text-decoration: none !important;
      }

      .card-cell a:hover {
        color: #6ad1e3;
      }
    `}</style>
  </Card>
);
