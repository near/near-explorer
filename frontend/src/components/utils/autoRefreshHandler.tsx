import InfiniteScroll from "react-infinite-scroll-component";

import React from "react";

import PaginationSpinner from "./PaginationSpinner";

interface Config {
  fetchDataFn: Function;
  count: number;
  category: string;
  dashboard?: boolean;
}

export default (
  WrappedComponent: React.ComponentType<any>,
  config: Config,
  props?: any
) => {
  return class extends React.Component {
    constructor(props: any) {
      super(props);
      this.timer = null;
    }

    timer: ReturnType<typeof setTimeout> | null;

    state = {
      items: Array<any>(),
      itemsLength: config.count,
      display: false,
      hasMore: true,
      loading: false
    };

    componentDidMount() {
      this.timer = setTimeout(this.regularFetchInfo, 0);
    }

    componentWillUnmount() {
      const timer = this.timer;
      this.timer = null;
      if (timer !== null) {
        clearTimeout(timer);
      }
    }

    componentDidUpdate(preProps: any) {
      if (this.props !== preProps) {
        this.setState({
          items: Array<any>(),
          itemsLength: config.count,
          display: false,
          hasMore: true,
          loading: false
        });
      }
    }

    regularFetchInfo = () => {
      if (config.dashboard) {
        this.setState({ hasMore: false });
      }
      if (this.state.itemsLength > 0) {
        this.fetchInfo(this.state.itemsLength);
        if (this.timer !== null) {
          this.timer = setTimeout(this.regularFetchInfo, 10000);
        }
      }
    };

    fetchInfo = (count: number) => {
      config
        .fetchDataFn(count)
        .then((items: any) => {
          let newState: any;
          if (items.length > 0) {
            newState = {
              items,
              itemsLength: items.length,
              display: true
            };
            if (items.length < config.count) {
              newState.hasMore = false;
            }
          } else {
            newState = {
              hasMore: false,
              display: true,
              itemsLength: 0
            };
          }
          this.setState(newState);
        })
        .catch((err: Error) => {
          console.error(err);
        });
    };

    fetchMoreData = async () => {
      this.setState({ loading: true });
      const endTimestamp = this.getEndTimestamp(config.category);
      config
        .fetchDataFn(config.count, endTimestamp)
        .then((newData: any) => {
          let newState: any;
          if (newData.length > 0) {
            const items = this.state.items.concat(newData);
            newState = {
              items,
              itemsLength: items.length,
              loading: false
            };
            if (newData.length < config.count) {
              newState.hasMore = false;
            }
          } else {
            newState = { hasMore: false, loading: false };
          }
          this.setState(newState);
        })
        .catch((err: Error) => console.error(err));
    };

    getEndTimestamp = (category: string) => {
      let endTimestamp;
      switch (category) {
        case "Account":
          endTimestamp = this.state.items[this.state.items.length - 1]
            .createdAtBlockTimestamp;
          break;
        case "Block":
          endTimestamp = this.state.items[this.state.items.length - 1]
            .timestamp;
          break;
        case "Node":
          endTimestamp = this.state.items[this.state.items.length - 1].lastSeen;
          break;
        case "Transaction":
          endTimestamp = this.state.items[this.state.items.length - 1]
            .blockTimestamp;
          break;
        default:
          endTimestamp = undefined;
      }
      return endTimestamp;
    };

    render() {
      if (!this.state.display) {
        return <PaginationSpinner hidden={false} />;
      }
      return (
        <>
          <InfiniteScroll
            dataLength={this.state.items.length}
            next={this.fetchMoreData}
            hasMore={this.state.hasMore}
            loader={
              this.state.loading ? (
                <PaginationSpinner hidden={false} />
              ) : (
                <>
                  <button
                    onClick={this.fetchMoreData}
                    className="load-button"
                    style={{ display: this.state.hasMore ? "block" : "none" }}
                  >
                    Load More
                  </button>
                </>
              )
            }
            style={{ overflowX: "hidden" }}
          >
            <WrappedComponent items={this.state.items} {...props} />
          </InfiniteScroll>
          <style jsx global>{`
            .load-button {
              width: 100px;
              background-color: #f8f8f8;
              display: block;
              text-align: center;
              text-decoration: none;
              font-family: BentonSans;
              font-size: 14px;
              color: #0072ce;
              font-weight: bold;
              text-transform: uppercase;
              margin: 20px auto;
              border-radius: 30px;
              padding: 8px 0;
              cursor: pointer;
              border: none;
            }
          `}</style>
        </>
      );
    }
  };
};
