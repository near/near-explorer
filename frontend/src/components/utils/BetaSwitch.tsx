import React from "react";

import { Switch } from "@/frontend/components/utils/Switch";
import { useBetaOptions } from "@/frontend/hooks/use-beta-options";
import { styled } from "@/frontend/libraries/styles";

const Wrapper = styled("div", {
  display: "flex",
  alignItems: "center",
  color: "#ccc",

  "& > * + *": {
    marginLeft: 8,
  },

  variants: {
    enabled: {
      true: {
        color: "#0072ce",
      },
    },
  },
});

export const BetaSwitch: React.FC = () => {
  const [betaOptions, setBetaOptions] = useBetaOptions();
  return (
    <Wrapper enabled={betaOptions?.enabled}>
      <span>Beta</span>
      <Switch
        checked={Boolean(betaOptions?.enabled)}
        onChange={(enabled) => setBetaOptions((prev) => ({ ...prev, enabled }))}
      />
    </Wrapper>
  );
};
