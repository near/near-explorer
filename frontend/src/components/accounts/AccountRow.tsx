import moment from "moment";
import Link from "next/link";

import React from "react";
import { Row, Col } from "react-bootstrap";

import AccountsApi from "../../libraries/explorer-wamp/accounts";

import Balance from "../utils/Balance";

export interface Props {
  accountId: string;
  createdAt: number;
}

export interface State {
  amount: string;
}

export default class extends React.Component<Props, State> {
  state: State = {
    amount: ""
  };

  _getDetail = async () => {
    const detail = await new AccountsApi().queryAccount(this.props.accountId);
    this.setState({ amount: detail.amount });
  };

  componentDidMount() {
    this._getDetail();
  }

  render() {
    const { accountId, createdAt } = this.props;
    const { amount } = this.state;
    return (
      <Link href="/accounts/[id]" as={`/accounts/${accountId}`}>
        <a style={{ textDecoration: "none" }}>
          <Row className="transaction-row mx-0">
            <Col md="auto" xs="1" className="pr-0">
              <img
                src="/static/images/icon-t-acct.svg"
                style={{ width: "15px" }}
              />
            </Col>
            <Col md="7" xs="6">
              <Row>
                <Col className="transaction-row-title">@{accountId}</Col>
              </Row>
            </Col>
            <Col md="3" xs="4" className="ml-auto text-right">
              <Row>
                <Col className="transaction-row-txid">
                  {amount && <Balance amount={amount} />}
                </Col>
              </Row>
              <Row>
                <Col className="transaction-row-timer">
                  {createdAt && `Created at ${moment(createdAt).format("LLL")}`}
                </Col>
              </Row>
            </Col>
            <style jsx global>{`
              .transaction-row {
                padding-top: 10px;
                padding-bottom: 10px;
                border-top: solid 2px #f8f8f8;
              }

              .transaction-row:hover {
                background: rgba(0, 0, 0, 0.1);
              }

              .transaction-row-title {
                font-family: BentonSans;
                font-size: 14px;
                font-weight: 500;
                line-height: 1.29;
                color: #24272a;
              }

              .transaction-row-text {
                font-family: BentonSans;
                font-size: 12px;
                font-weight: 500;
                line-height: 1.5;
                color: #999999;
              }

              .transaction-row-txid {
                font-family: BentonSans;
                font-size: 14px;
                font-weight: 500;
                line-height: 1.29;
                color: #0072ce;
              }

              .transaction-row-timer {
                font-family: BentonSans;
                font-size: 12px;
                color: #999999;
                font-weight: 100;
              }

              .transaction-row-timer-status {
                font-weight: 500;
              }
            `}</style>
          </Row>
        </a>
      </Link>
    );
  }
}
