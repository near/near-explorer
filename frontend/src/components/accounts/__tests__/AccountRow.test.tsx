import { renderElement } from "../../../testing/utils";

import AccountRow from "../AccountRow";

describe("<AccountRow />", () => {
  beforeEach(() => jest.resetAllMocks());

  it("renders with short name", () => {
    expect(renderElement(<AccountRow accountId="account" />)).toMatchSnapshot();
  });

  it("renders with long name", () => {
    expect(
      renderElement(
        <AccountRow accountId="b7df2090560a225dc4934aed43db03a6c674c2d4.lockup.near" />
      )
    ).toMatchSnapshot();
  });
});
