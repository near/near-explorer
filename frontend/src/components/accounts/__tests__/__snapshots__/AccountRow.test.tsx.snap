// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccountRow /> renders with long name 1`] = `
<a
  className="c-StyledLink-hIXuvg"
  href="/accounts/b7df2090560a225dc4934aed43db03a6c674c2d4.lockup.near"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="c-TransactionRow-keHKEY mx-0 row"
  >
    <div
      className="pr-0 col-md-auto col-1"
    >
      <img
        className="c-AccountIcon-exgREZ"
        src="/static/images/icon-t-acct.svg"
      />
    </div>
    <div
      className="c-TransactionRowTitle-bvTgKA pt-1 col-md-7 col-11"
    >
      b7df2090560a225dc4934aed43db03a6c674c2d4.lockup.near
    </div>
    <div
      className="c-TransactionRowTransactionId-lbSlCc ml-auto pt-1 text-right col-md-3 col-5"
    >
      <div
        className="spinner-border"
      />
      <div
        className="c-TransactionRowTimer-ehtocQ"
      >
        component.accounts.AccountRow.created_on
         
        2020-09-30T10:23:09Z formatted by PPP
        <div
          className="c-Wrapper-kgHvMp c-Wrapper-kgHvMp-ieesTsS-css"
        >
          <svg
            height=".6em"
            onClick={[Function]}
            viewBox="0 0 18 18"
            width=".6em"
          >
            <path
              d="m11.315 5.9932h-9.7723c-0.85717 0-1.5427 0.63916-1.5427 1.4383v9.1302c0 0.7788 0.68557 1.4383 1.5427 1.4383h9.7931c0.8354 0 1.5427-0.6392 1.5427-1.4383l8e-4 -9.1302c-0.0217-0.79915-0.7073-1.4383-1.5645-1.4383h2e-4zm0.2569 10.549c0 0.1201-0.1071 0.22-0.236 0.22l-9.793-7e-4c-0.12891 0-0.23606-0.0999-0.23606-0.2201l8.3e-4 -9.11c0-0.12018 0.10715-0.22009 0.23607-0.22009h9.793c0.1289 0 0.2361 0.0999 0.2361 0.22009l-9e-4 9.1108z"
              fill="currentColor"
            />
            <path
              d="m16.457 1.1984e-4h-9.7723c-0.85718 0-1.5427 0.63916-1.5427 1.4383v3.3566h1.2858v-3.3566c0-0.12019 0.10714-0.22009 0.23606-0.22009h9.793c0.1289 0 0.2361 0.09989 0.2361 0.22009v9.1302c0 0.1202-0.1071 0.2201-0.2361 0.2201h-2.3146v1.1987h2.3146c0.8354 0 1.5428-0.6392 1.5428-1.4383v-9.1108c0-0.79916-0.6856-1.4383-1.5428-1.4383l2e-4 1.1984e-4z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</a>
`;

exports[`<AccountRow /> renders with short name 1`] = `
<a
  className="c-StyledLink-hIXuvg"
  href="/accounts/account"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="c-TransactionRow-keHKEY mx-0 row"
  >
    <div
      className="pr-0 col-md-auto col-1"
    >
      <img
        className="c-AccountIcon-exgREZ"
        src="/static/images/icon-t-acct.svg"
      />
    </div>
    <div
      className="c-TransactionRowTitle-bvTgKA pt-1 col-md-7 col-11"
    >
      account
    </div>
    <div
      className="c-TransactionRowTransactionId-lbSlCc ml-auto pt-1 text-right col-md-3 col-5"
    >
      <div
        className="spinner-border"
      />
      <div
        className="c-TransactionRowTimer-ehtocQ"
      >
        component.accounts.AccountRow.created_on
         
        2020-09-30T10:23:09Z formatted by PPP
        <div
          className="c-Wrapper-kgHvMp c-Wrapper-kgHvMp-ieesTsS-css"
        >
          <svg
            height=".6em"
            onClick={[Function]}
            viewBox="0 0 18 18"
            width=".6em"
          >
            <path
              d="m11.315 5.9932h-9.7723c-0.85717 0-1.5427 0.63916-1.5427 1.4383v9.1302c0 0.7788 0.68557 1.4383 1.5427 1.4383h9.7931c0.8354 0 1.5427-0.6392 1.5427-1.4383l8e-4 -9.1302c-0.0217-0.79915-0.7073-1.4383-1.5645-1.4383h2e-4zm0.2569 10.549c0 0.1201-0.1071 0.22-0.236 0.22l-9.793-7e-4c-0.12891 0-0.23606-0.0999-0.23606-0.2201l8.3e-4 -9.11c0-0.12018 0.10715-0.22009 0.23607-0.22009h9.793c0.1289 0 0.2361 0.0999 0.2361 0.22009l-9e-4 9.1108z"
              fill="currentColor"
            />
            <path
              d="m16.457 1.1984e-4h-9.7723c-0.85718 0-1.5427 0.63916-1.5427 1.4383v3.3566h1.2858v-3.3566c0-0.12019 0.10714-0.22009 0.23606-0.22009h9.793c0.1289 0 0.2361 0.09989 0.2361 0.22009v9.1302c0 0.1202-0.1071 0.2201-0.2361 0.2201h-2.3146v1.1987h2.3146c0.8354 0 1.5428-0.6392 1.5428-1.4383v-9.1108c0-0.79916-0.6856-1.4383-1.5428-1.4383l2e-4 1.1984e-4z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</a>
`;
